core/node/node_framework/src/service/mod.rs

```diff
@@ -187,37 +177,46 @@ impl ZkStackService {
             tracing::info!("Remaining tasks finished without reaching timeouts");
         }
 
-        // Call after_node_shutdown hooks.
-        let local_set = tokio::task::LocalSet::new();
-        let join_handles = tasks.iter_mut().filter_map(|task| {
-            task.after_node_shutdown
-                .take()
-                .map(|task| local_set.spawn_local(task))
-        });
-        local_set.block_on(&self.runtime, futures::future::join_all(join_handles));
-
-        if failure {
-            anyhow::bail!("Task {resolved_task_name} failed");
-        } else {
-            Ok(())
-        }
-    }
-
-    pub(crate) fn stop_receiver(&self) -> StopReceiver {
-        StopReceiver(self.stop_sender.subscribe())
+        result
     }
 }
 
-struct TaskRepr {
-    name: String,
-    task: Option<BoxFuture<'static, anyhow::Result<()>>>,
-    after_node_shutdown: Option<BoxFuture<'static, ()>>,
-}
+fn oneshot_runner_task(
+    oneshot_tasks: Vec<BoxFuture<'static, anyhow::Result<()>>>,
+    mut stop_receiver: StopReceiver,
+    only_oneshot_tasks: bool,
+) -> BoxFuture<'static, anyhow::Result<()>> {
+    Box::pin(async move {
+        let oneshot_tasks = oneshot_tasks.into_iter().map(|fut| async move {
+            // Spawn each oneshot task as a separate tokio task.
+            // This way we can handle the cases when such a task panics and propagate the message
+            // to the service.
+            let handle = tokio::runtime::Handle::current();
+            match handle.spawn(fut).await {
+                Ok(Ok(())) => Ok(()),
+                Ok(Err(err)) => Err(err),
+                Err(panic_err) => {
+                    let panic_msg = try_extract_panic_message(panic_err);
+                    Err(anyhow::format_err!("Precondition panicked: {panic_msg}"))
```

Comment:
Nit: Misleading error message; AFAIU, the run tasks are not necessarily preconditions (maybe, it could make sense to pass around the task kind to make messages like this more informative?).

Reply:
Thanks, leftover from times when this code was only meant for preconditions.
Passing the task type maybe makes sense, but I think for the 1st iteration it should be sufficient that the panic backtrace will anyway be available (and for prod environment it'll anyway be the thing that we see first), so I'll leave it mostly as is.